perses/dispersed/feptasks.py

```diff
@@ -84,30 +111,103 @@ def _get_integrator(self, thermodynamic_state):
         """
         return self._integrator
 
-    def reset(self):
-        """
-        Reset the work statistics on the associated ContextCache integrator.
+    def apply(self, thermodynamic_state, sampler_state):
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It will apply the full NCMC protocol.
 
         Parameters
         ----------
         thermodynamic_state : openmmtools.states.ThermodynamicState
-            the thermodynamic state for which this integrator is cached.
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
         """
-        self._integrator.reset()
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It also logs
+        benchmarking information through the utils.Timer class.
+        Parameters
+        ----------
+        thermodynamic_state : openmmtools.states.ThermodynamicState
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
+        See Also
+        --------
+        openmmtools.utils.Timer
+        """
+        # Check if we have to use the global cache.
+        if self.context_cache is None:
+            context_cache = cache.global_context_cache
+        else:
+            context_cache = self.context_cache
+
+        # Create integrator.
+        integrator = self._get_integrator(thermodynamic_state)
+
+        context, integrator = context_cache.get_context(thermodynamic_state, integrator)
```

Comment:
Checking just for caution: make sure the `integrator` here has always the same serialization hash every time you call apply (i.e. all the variables that are not standardized must be the same everytime you call apply), or the `ContextCache` won't be able to match it to an existing `Context`.

We can expand the list of the serialized attributes if you need it (see here: https://github.com/choderalab/openmmtools/blob/master/openmmtools/cache.py#L438), or you can simply add your attributes to that set at runtime.

Reply:
I'm not sure I'm clear on this one. `apply` should only be called once in each protocol, and a copy of the same integrator is sent each time, so I think we're ok? Or is the serialization going to mess that up?

Reply:
Can you protect this so that subsequent calls to `apply` do nothing?

Reply:
It may be fine then! I just wanted to make sure you were aware of this problem in case you started to see that your `Context`s are not recycled properly.

Reply:
>Can you protect this so that subsequent calls to apply do nothing?

Yes. Going forward, is this adherent with the design principles of the `MCMCMove`? What's the best way to deal with this situation that doesn't result in an `MCMCMove` that behaves anomalously?